man/tpm2_flushcontext.1.md
% tpm2_flushcontext(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_flushcontext**(1) - Remove a specified handle, or all contexts associated
with a transient object, loaded session or saved session from the TPM.

# SYNOPSIS

**tpm2_flushcontext** [*OPTIONS*] [*ARGUMENT*]

# DESCRIPTION

**tpm2_flushcontext**(1) - Remove a specified handle, or all contexts associated
with a transient object, loaded session or saved session from the TPM. The
object to be flushed is specified as the first argument to the tool and is in
one of the following forms:
  - The handle of the object to be flushed from the TPM. Must be a valid handle
  number.
  - Flush a session via a session file. A session file is generated from
    **tpm2_startauthsession**(1)'s **-S** option.

# OPTIONS

  * **-t**, **\--transient-object**:

    Remove all transient objects.

  * **-l**, **\--loaded-session**:

    Remove all loaded sessions.

  * **-s**, **\--saved-session**:

    Remove all saved sessions.

  * **ARGUMENT** the command line argument specifies the _OBJECT_ to be removed
    from the TPM resident memory.

  * **\--cphash**=_FILE_

    File path to record the hash of the command parameters. This is commonly
    termed as cpHash. NOTE: When this option is selected, The tool will not
    actually execute the command, it simply returns a cpHash.

[common options](common/options.md)

[common tcti options](common/tcti.md)

# EXAMPLES

## Flushing a Transient Object

Typically, when using the TPM, the interactions occur through a resource
manager, like tpm2-abrmd(8). When the process exits, transient object handles
are flushed. Thus, flushing transient objects through the command line is not
required. However, when interacting with the TPM directly, this scenario is
possible. The below example assumes direct TPM access not brokered by a resource
manager. Specifically we will use the simulator.

```bash
tpm2_createprimary -Tmssim -c primary.ctx

tpm2_getcap -T mssim handles-transient
- 0x80000000

tpm2_flushcontext -T mssim 0x80000000
```

## Flush All the Transient Objects
```bash
tpm2_flushcontext \--transient-object
```

## Flush a Session
```bash
tpm2_startauthsession -S session.dat

tpm2_flushcontext session.dat
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_flushcontext.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdlib.h>

#include "files.h"
#include "log.h"
#include "object.h"
#include "tpm2.h"
#include "tpm2_tool.h"
#include "tpm2_capability.h"
#include "tpm2_options.h"
#include "tpm2_session.h"

#define MAX_SESSIONS 3
#define TOTAL_CTX_TYPES 3 // contexts are transient/ loaded/ saved.
#define MAX_CTX_COUNT 255
struct tpm_flush_context_ctx {
    /*
     * Inputs
     */
    unsigned encountered_option_flags;
    TPM2_HANDLE property[TOTAL_CTX_TYPES];
    ESYS_TR context_handles[MAX_CTX_COUNT]; //ESYS_TR
    uint8_t context_handle_count;
    bool is_t_l_s_specified; //t l s option combination

    bool is_arg_session;
    bool is_arg_transient;
    char *context_arg;
    tpm2_session *arg_session;

    /*
     * Outputs
     */

    /*
     * Parameter hashes
     */
    const char *cp_hash_path;
    TPM2B_DIGEST cp_hash;
    bool is_command_dispatch;
    TPMI_ALG_HASH parameter_hash_algorithm;

};

static struct tpm_flush_context_ctx ctx = {
    .context_handles = {ESYS_TR_NONE},
    .parameter_hash_algorithm = TPM2_ALG_ERROR
};

static const char *get_property_name(TPM2_HANDLE handle) {

    switch (handle & TPM2_HR_RANGE_MASK) {
    case TPM2_HR_TRANSIENT:
        return "transient";
    case TPM2_HT_LOADED_SESSION << TPM2_HR_SHIFT:
        return "loaded session";
    case TPM2_HT_SAVED_SESSION << TPM2_HR_SHIFT:
        return "saved session";
    }

    return "invalid";
}

static tool_rc flushcontext(ESYS_CONTEXT *ectx) {

    tool_rc rc = tool_rc_success;
    tool_rc tmp_rc = tool_rc_success;
    uint32_t i;
    for (i = 0; i < ctx.context_handle_count; ++i) {
        /*
         * Continue flushing the handles after error AND
         * Capture the error as final return data.
         */
        tmp_rc = tpm2_flush_context(ectx, ctx.context_handles[i], &ctx.cp_hash,
            ctx.parameter_hash_algorithm);
        if (tmp_rc != tool_rc_success) {
            LOG_ERR("Failed Flush Context for %s handle 0x%x",
                get_property_name(ctx.context_handles[i]),
                ctx.context_handles[i]);
            rc = tmp_rc;
        }
    }

    return rc;
}

static tool_rc process_outputs(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Outputs that do not require TPM2_CC_<command> dispatch
     */
    bool is_file_op_success = true;
    if (ctx.cp_hash_path) {
        is_file_op_success = files_save_digest(&ctx.cp_hash, ctx.cp_hash_path);

        if (!is_file_op_success) {
            return tool_rc_general_error;
        }
    }

    tool_rc rc = tool_rc_success;
    if (!ctx.is_command_dispatch) {
        return rc;
    }

    /*
     * 2. Outputs generated after TPM2_CC_<command> dispatch
     */

    return tool_rc_success;
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);
    /*
     * 1. Object and auth initializations
     */

    /*
     * 1.a Add the new-auth values to be set for the object.
     */

    /*
     * 1.b Add object names and their auth sessions
     */
    tool_rc rc = tool_rc_success;
    TPM2_HANDLE sapi_handle = 0;
    if (!ctx.is_t_l_s_specified) {
        ctx.is_arg_transient = tpm2_util_string_to_uint32(ctx.context_arg,
            &sapi_handle);
    }

    if (ctx.is_arg_transient) {
        rc = tpm2_util_sys_handle_to_esys_handle(ectx, sapi_handle,
            ctx.context_handles);
        if (rc != tool_rc_success) {
            LOG_ERR("Handle not found.");
            return rc;
        }

        ctx.context_handle_count++;
    }
    
    if (!ctx.is_t_l_s_specified && !ctx.is_arg_transient) {
        rc = tpm2_session_restore(ectx, ctx.context_arg, true,
            &ctx.arg_session);
        if (rc == tool_rc_success) {
            ctx.is_arg_session = true;
            ctx.context_handles[0] = tpm2_session_get_handle(ctx.arg_session);
            ctx.context_handle_count++;
        }
    }

    if (!ctx.is_t_l_s_specified && !ctx.is_arg_transient &&
    !ctx.is_arg_session) {
        LOG_ERR("Argument neither a session nor a transient.");
        return tool_rc_general_error;
    }

    /*
     * 2. Restore auxiliary sessions
     */

    /*
     * 3. Command specific initializations
     */

    /*
     * Populate ctx.context_handles with transient, loaded and saved handles
     * Note: encountered_option is nil when context is specified as argument.
     */
    TPM2_HANDLE property[3] = { 0 };
    unsigned offset = 0;
    if (ctx.encountered_option_flags & 1 << 0) {
        property[offset++] = TPM2_TRANSIENT_FIRST;
    }
    if (ctx.encountered_option_flags & 1 << 1) {
        property[offset++] = TPM2_LOADED_SESSION_FIRST;
    }
    if (ctx.encountered_option_flags & 1 << 2) {
        property[offset++] = TPM2_ACTIVE_SESSION_FIRST;
    }

    unsigned i = 0; // Iterates through t,l,s types
    for (i = 0; i < offset; i++) {
        TPM2_HANDLE p = property[i];
        TPMS_CAPABILITY_DATA *capability_data;
        rc = tpm2_capability_get(ectx, TPM2_CAP_HANDLES, p,
            TPM2_MAX_CAP_HANDLES, &capability_data);
        if (rc != tool_rc_success) {
            LOG_ERR("Error reading handle info from TPM.");
            return tool_rc_general_error;
        }
    
        unsigned j = 0; //Iterates through all available handles in t/l/s
        for (j = 0; j < capability_data->data.handles.count; j++) {
            rc = tpm2_util_sys_handle_to_esys_handle(ectx,
                capability_data->data.handles.handle[j],
                &ctx.context_handles[j]);
            if (rc != tool_rc_success) {
                LOG_ERR("Error reading handle info from TPM.");
                return tool_rc_general_error;
            }
            ctx.context_handle_count++;
        }

        free(capability_data);
    }

    /*
     * 4. Configuration for calculating the pHash
     */

    /*
     * 4.a Determine pHash length and alg
     */
    tpm2_session *all_sessions[MAX_SESSIONS] = {
        0,
        0,
        0
    };

    const char **cphash_path = ctx.cp_hash_path ? &ctx.cp_hash_path : 0;

    ctx.parameter_hash_algorithm = tpm2_util_calculate_phash_algorithm(ectx,
        cphash_path, &ctx.cp_hash, 0, 0, all_sessions);

    /*
     * 4.b Determine if TPM2_CC_<command> is to be dispatched
     */
    ctx.is_command_dispatch = ctx.cp_hash_path ? false : true;

    /*
     * Note: pHash cannot be calculated if multiple sessions are found needing
     *       to be flushed. This is because pHash output needs to be written to
     *       a single file which must contain a single pHash.
     * 
     *       This implies if ctx.context_handle_count > 1 and pHash must be
     *       be calculated --> Error out.
     */

    return tool_rc_success;
}

static tool_rc check_options(void) {

    /*
     * Either an argument specifying the context to flush
     * OR
     * Option to flush all must be specified
     */
    ctx.is_t_l_s_specified = ctx.encountered_option_flags;
    if (!ctx.is_t_l_s_specified && !ctx.context_arg) {
        LOG_ERR("Specify options to evict handles or a session context.");
        return tool_rc_option_error;
    }

    /*
     * Options are mutually exclusive of an argument
     */
    if (ctx.is_t_l_s_specified && ctx.context_arg) {
        LOG_ERR("Specify either 't' 'l' 's' or a context. Cannot specify both");
        return tool_rc_option_error;
    }

    return tool_rc_success;
}

static bool on_arg(int argc, char *argv[]) {

    if (argc > 1) {
        LOG_ERR("Specify one context.");
        return tool_rc_option_error;
    }

    ctx.context_arg = argv[0];

    return true;
}

static bool on_option(char key, char *value) {

    UNUSED(value);

    switch (key) {
    case 't':
        ctx.encountered_option_flags |= 1 << 0;
        break;
    case 'l':
        ctx.encountered_option_flags |= 1 << 1;
        break;
    case 's':
        ctx.encountered_option_flags |= 1 << 2;
        break;
    case 0:
        ctx.cp_hash_path = value;
        break;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static const struct option topts[] = {
        { "transient-object", no_argument, 0, 't' },
        { "loaded-session",   no_argument, 0, 'l' },
        { "saved-session",    no_argument, 0, 's' },
        { "cphash",     required_argument, 0,  0  },
    };

    *opts = tpm2_options_new("tls", ARRAY_LEN(topts), topts, on_option, on_arg,
            0);

    return *opts != 0;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    /*
     * 1. Process options
     */
    tool_rc rc = check_options();
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 2. Process inputs
     */
    rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 3. TPM2_CC_<command> call
     */
    rc = flushcontext(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 4. Process outputs
     */
    return process_outputs(ectx);
}

static tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Free objects
     */
    if (ctx.is_arg_session) {
        tpm2_session_free(&ctx.arg_session);
    }

    /*
     * 2. Close authorization sessions
     */

    /*
     * 3. Close auxiliary sessions
     */

    return tool_rc_success;
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("flushcontext", tpm2_tool_onstart, tpm2_tool_onrun,
    tpm2_tool_onstop, 0)
